exporter/clickhouselogsexporter/exporter.go

```diff
@@ -0,0 +1,284 @@
+		return nil, err
+	}
+
+	client, err := newClickhouseClient(logger, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	insertLogsSQL := renderInsertLogsSQL(cfg)
+
+	return &clickhouseLogsExporter{
+		db:            client,
+		insertLogsSQL: insertLogsSQL,
+		logger:        logger,
+		cfg:           cfg,
+	}, nil
+}
+
+// Shutdown will shutdown the exporter.
+func (e *clickhouseLogsExporter) Shutdown(_ context.Context) error {
+	if e.db != nil {
+		return e.db.Close()
+	}
+	return nil
+}
+
+func (e *clickhouseLogsExporter) pushLogsData(ctx context.Context, ld plog.Logs) error {
+	start := time.Now()
+	statement, err := e.db.PrepareBatch(ctx, e.insertLogsSQL)
+	if err != nil {
+		return fmt.Errorf("PrepareBatch:%w", err)
+	}
+	defer func() {
+		_ = statement.Abort()
+	}()
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		logs := ld.ResourceLogs().At(i)
+		res := logs.Resource()
+		resources := attributesToSlice(res.Attributes())
+		for j := 0; j < logs.ScopeLogs().Len(); j++ {
+			rs := logs.ScopeLogs().At(j).LogRecords()
+			for k := 0; k < rs.Len(); k++ {
+				r := rs.At(k)
+
+				id, err := ksuid.NewRandomWithTime(r.Timestamp().AsTime())
+				if err != nil {
+					return fmt.Errorf("error creating id: %w", err)
+				}
+
+				attributes := attributesToSlice(r.Attributes())
+				err = statement.Append(
+					uint64(r.Timestamp()),
+					uint64(r.ObservedTimestamp()),
+					id.String(),
+					r.TraceID().HexString(),
+					r.SpanID().HexString(),
+					r.Flags(),
+					r.SeverityText(),
+					int32(r.SeverityNumber()),
+					r.Body().AsString(),
+					resources.StringKeys,
+					resources.StringValues,
+					attributes.StringKeys,
+					attributes.StringValues,
+					attributes.IntKeys,
+					attributes.IntValues,
+					attributes.FloatKeys,
+					attributes.FloatValues,
+				)
+				if err != nil {
+					return fmt.Errorf("StatementAppend:%w", err)
+				}
+
+			}
+		}
+	}
+	err = statement.Send()
+	if err != nil {
+		return fmt.Errorf("StatementSend:%w", err)
+	}
+	duration := time.Since(start)
+	e.logger.Debug("insert logs", zap.Int("records", ld.LogRecordCount()),
+		zap.String("cost", duration.String()))
+	return err
+}
+
+type attributesToSliceResponse struct {
+	StringKeys   []string
+	StringValues []string
+	IntKeys      []string
+	IntValues    []int64
+	FloatKeys    []string
+	FloatValues  []float64
+}
+
+func attributesToSlice(attributes pcommon.Map) (response attributesToSliceResponse) {
+	attributes.Range(func(k string, v pcommon.Value) bool {
+		switch v.Type().String() {
+		case "INT":
+			response.IntKeys = append(response.IntKeys, formatKey(k))
+			response.IntValues = append(response.IntValues, v.IntVal())
+		case "DOUBLE":
+			response.FloatKeys = append(response.FloatKeys, formatKey(k))
+			response.FloatValues = append(response.FloatValues, v.DoubleVal())
+		default: // store it as string
+			response.StringKeys = append(response.StringKeys, formatKey(k))
+			response.StringValues = append(response.StringValues, v.AsString())
```

Comment:
we won't allow storing nested. If user want's to search on latest they will have to use `json_parser` in otel_config to parse it and create an attribute out of it.

Reply:
>we won't allow storing nested

Where are we preventing this from happening?

Reply:
here https://github.com/SigNoz/signoz-otel-collector/blob/c52666819bc5e1ab227103315f24c3bcebb67089/exporter/clickhouselogsexporter/exporter.go#L139

Reply:
>they will have to use json_parser in otel_config to parse it and create an attribute out of it.

By that did you mean they have to flatten and create new attribute for primitive types?

>we won't allow storing nested

This means if the type is Map or Slice it should drop them and log a message? But `attributesToSlice` This func has default for everything other than `INT` and `DOUBLE`. The `asString` returns the json encoded value for slice and map.

Reply:
Yes, and in the future, we will have an option in the UI to parse attributes. I am not sure if we should drop them. As we can clearly mention in the docs that it will be converted to a string ( by this way they will at least see the data and change the config accordingly)

Reply:
>we will have an option in the UI to parse attributes

Can you explain this in more detail? 

>can clearly mention in the docs that it will be converted to a string

Can you point me to this doc?

I am confused since you said we don't allow storing nested but this indeed does stores the nested but serialized data. Is it _we won't allow storing nested_ or _we won't allow searching on nested_? If the data is stored and searching is also allowed then It wouldn't be useful because I can't do search by using `.` or `_` separated key values?

Reply:
We don't have docs as of now.

To clarify the statement:- 
It's "we won't store nested as nested data type, we will store it as a string, if you want to parse nested values and convert them to individual attributes you can parse it via the `json_parser` operator".

Reply:
Ah, that's more clear. How can I search on this data? For instance take this as example.
```json
{
  "db_response": {
     "row": 12,
     "time_taken": 200 
   }
}
```

Will I be able to search based on `db_response.time_taken`?

Reply:
Hmm that's a bummer because after certain point it's not practical for operators to collect the keys from devs and update the config. If I update the config in the later days does it backfill the values for past data?

Reply:
Yea that's correct, If the config is updated it won't backfill it will just parse the new logs.


That's why in the next version we will have the option to parse these from the Signoz UI. The flow might change but it will look something like `User will select the field that they want to search/parse independently and we will suggest regex to parse it and user can edit it as required and then based on it we will create a materialized column`.

Reply:
in JSON If the number of keys explodes, the creation of new keys becomes very sow and the ingestion decreases significantly since it was still in the experimental phase, we decided best to avoid it as of now.